Unread @-mentions indicators for topic

SilentCruzer opened the pull request:
Closes: #4857

Changes made:

- [Commit:- unreadSelectors: Add selector to return ids of unread mentions](https://github.com/SilentCruzer/zulip-mobile/commit/45b76f13a6ae3c95bdcb602c29ebd065cc7dee0a)
     - Created a new selector to return state.unread.mentions.

- [Commit:- unreadSelectors: Push msgIds in type UnreadStreamItem](https://github.com/SilentCruzer/zulip-mobile/commit/a756c24c58bb20924edddd68912b9fe8a1814e0b) 
    - Modified the function getUnreadStreamsAndTopics to return the array of ids of unread messages.

- [Commit:- unreadCards: Add indicator for topics with unread mentions](https://github.com/SilentCruzer/zulip-mobile/commit/255e1b49e516045403871cfef2b9251b0d51aefb)
     - Created method getMentions which returns the number of unread messages per topic.
     - Added the indicator to the left side of the number of unread messages.

Screenshots: 
<img src="https://user-images.githubusercontent.com/73246484/156561841-6420f3c0-3a60-469b-a239-d7d2a65b1dd1.jpg" width=250 height=600>     <img src="https://user-images.githubusercontent.com/73246484/156561890-b9060505-bf31-499f-9ec2-32158b507417.jpg" width=250 height=600>

gnprice commented on src/unread/UnreadCards.js:
```diff
@@ -62,6 +68,15 @@ export default function UnreadCards(props: Props): Node {
+  // This method return the number of unread mentions per topic
+  const getMentions = (stream_id: number, topic: string) => {
+    let unreadMentions = 0;
+    const streamData = unreadStreamsAndTopics.find(item => item.streamId === stream_id);
+    const topicData = streamData?.data.find(item => item.topic === topic);
```
Searches like these are inefficient; see #3339. We have some in the existing code, but shouldn't add more.

Here, this is quite inefficient overall because for every one of the `TopicItem` components, we're going to end up searching through the whole list of unread streams and the whole list of unread topics in the relevant stream. So that's a loop of loops, and takes quadratic time.

SilentCruzer replied:
Thanks, I now realized the mistake I made. I was trying to categorize how many mentions each topic had.

So before I make the changes, I just want to understand the suggested [implementation](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/unread.20mentions/near/1337465). In this, it is mentioned to compare the unread mentions with the local msgs.  So when we return the data of the topic, we can also return a boolean (if the user is mentioned or not) after comparing the msgIds with the unread mentions. To return it, we can use the data attribute in the variable named `total`.Does this sound right?

https://github.com/zulip/zulip-mobile/blob/255e1b49e516045403871cfef2b9251b0d51aefb/src/unread/UnreadCards.js#L29-L47

gnprice replied:
> So when we return the data of the topic, we can also return a boolean (if the user is mentioned or not) after comparing the msgIds with the unread mentions. To return it, we can use the data attribute in the variable named `total`.Does this sound right?

Yep, that's the right spot.


> Thanks, I now realized the mistake I made. I was trying to categorize how many mentions each topic had.

This is actually unrelated to the inefficiency here. If we did want to count the number of mentions, we would still want to do so inside the `getUnreadStreamsAndTopics` selector (so that the computation doesn't get repeated if the component rerenders for some unrelated reason), and would still want to do so in a way that looked at each piece of data just once so that it runs in linear time (aka O(n) time.)

For example, your current draft has:
```js
        let isMentioned = false;
        msgIds.forEach(item => {
          if (unreadMsgIds.has(item)) {
            isMentioned = true;
          }
        });
```

If we wanted the number of mentions instead of a boolean saying whether there are any mentions, we could change that to:
```js
        let numMentioned = 0;
        msgIds.forEach(item => {
          if (unreadMsgIds.has(item)) {
            numMentioned += 1;
          }
        });
```
and that would be just as efficient as the current draft.

gnprice reviewed (commented):
Thanks @SilentCruzer!

For the design, I think the intended design was to have just a red dot to say there were some @-mentions; not a count of how many there are. @alya, does that sound right to you?

For the implementation strategy for computing the data that drives this feature, please reread my description here:
https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/unread.20mentions/near/1337465
and try doing it that way. That will avoid the inefficiency described below.

alya commented:
> For the design, I think the intended design was to have just a red dot to say there were some @-mentions; not a count of how many there are. @alya, does that sound right to you?

Yeah, that sounds right. We should also consider showing an @-sign if that looks reasonable, as that's the design proposed by @terpimost for the equivalent webapp feature (https://github.com/zulip/zulip/issues/21637).

terpimost commented:
here is the design ref:
<img width="577" alt="image" src="https://user-images.githubusercontent.com/1903309/163681129-2ecb100c-f264-469b-9b9c-e3f18996762c.png">
<img width="573" alt="image" src="https://user-images.githubusercontent.com/1903309/163681158-b82f9bfc-9b64-481a-9bf7-0199c4359130.png">
Icon is `at-sign` from https://feathericons.com/?query=at

SilentCruzer commented:
@gnprice @alya, I have made the suggested changes, also changed the UI, and replaced the count with the @ sign.

Screenshots: 

 <img src="https://user-images.githubusercontent.com/73246484/164036492-385a27b1-b6e9-48a9-a765-8c8a618e0ad2.png" width=275 height=600>  <img src="https://user-images.githubusercontent.com/73246484/164036424-20796639-ac3a-4cbe-aba9-38d29d89272a.png" width=275 height=600>

gnprice commented on src/unread/unreadSelectors.js:
```diff
@@ -155,11 +157,17 @@ export const getUnreadStreamsAndTopics: Selector<$ReadOnlyArray<UnreadStreamItem
           total.unread += msgIds.size;
         }
-
+        let isMentioned = false;
+        msgIds.forEach(item => {
+          if (unreadMsgIds.has(item)) {
+            isMentioned = true;
+          }
+        });
```
This can be simplified in a couple of ways:
* After we've found one mention, there's no need to look through the rest of the list -- it won't change anything. So it'd be better to break out of the loop there.
  * The `Array#forEach` method doesn't support that. That's one of the reasons it's usually better to use a plain JS for-loop instead; e.g. `for (const item of msgIds) {` rather than `msgIds.forEach(item => {`. Then a `break` statement will end the loop.
* For this use case, there's an `Array` method to do it exactly: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some

So:
```suggestion
        const isMentioned = msgIds.some(id => unreadMsgIds.has(id));
```

Then at that point we may as well write it inline in the object literal below, instead of making a variable for it.

gnprice reviewed (commented):
Thanks @SilentCruzer for the revision!

The code looks good now, except a small comment below.

Oh, and a nit in the commit message:

    unreadCards: Add indicator for topics with unread mentions

That prefix "unreadCards" looks like it's an identifier, given its internal capitalization. But there is no identifier `unreadCards` in our codebase. A good prefix would be `UnreadCards`, which is the actual identifier that's the name of the component.

---

Looking at the design: I think there's a bit more space between the "@" and the unread count in your screenshots than there are in Vlad's design reference above. Can you narrow the spacing to match?

Also a lot of the other icons in your screenshot seem to be random other icons instead. That'd be good to debug over on chat.zulip.org. I don't think it's related to your change in this PR.

SilentCruzer commented:
Thanks for the review, I have made the suggested changes: 
 <img src="https://user-images.githubusercontent.com/73246484/164254667-b9e77c8b-8c3c-43c0-b4dd-fe26dc0a4230.png" width=275 height=600>  <img src="https://user-images.githubusercontent.com/73246484/164254303-ed1e6393-1a8e-4673-9248-7a6f64788f9c.png" width=275 height=600>

gnprice commented:
Thanks for the revision! All now looks good -- merging.

Oh, with a fix for one item in the commit message: it should mention the issue this fixes, in addition to doing so in the PR. See style guide:
https://github.com/zulip/zulip-mobile/blob/main/docs/style.md#mention-fixed-issue
